### ISSUE TITLE
anonymous users will be allowed to create networks

### ISSUE THREAD

**Issue by zufanka** [tenders-exposed/elvis-ember#167]
anonymous users will be allowed to create networks
These networks will be deleted as soon as the user closes the browser window

**Comment by georgiana-b** [comment]
We could just delete these userless networks on backend every week (for example) while ensuring that we have at least 10 of them at any given moment.
This way we don't have to worry about determining when the user closes the browser window and we can have a "latest networks page" that will feature the latest 10 networks thus solving #168.

**Comment by georgiana-b** [comment]
Of course, also have a disclaimer that tells users these public networks are public, so anybody can see, delete and update them. If they want to ensure persistency and consistency they have to log in.

If we apply my suggestion above, we should also tell them their networks will be in the open for everyone to see. This is another good incentive to log in, especially for journalists that want to keep their insights private.

**Comment by nightsh** [comment]
Uhm, the networks are already public. We probably need to change that as soon as we make this thing here happen.

**Comment by georgiana-b** [comment]
Yes, but if they will be featured on the "latest networks page" they will be much easier to find than if they were private networks. Unless the owner shares a private network it is pretty hard to guess its ID.

**Comment by nightsh** [comment]
You're so much right on this issue. That's just my opinion, but you have my upvotes.

I'd also say "Latest networks" to be actually "Latest anonymous networks" to avoid confusion, or anything explicit like that. Sounds like a nice idea.

**Comment by georgiana-b** [comment]
Considering the main motivation behind this is for people to explore networks without going through the hassle of making an account, I think we might end up not needing this if we implement third-party log in (Facebook, Twitter, GitHub) in the next backend which would remove that hassle. So, I would suggest that we reevaluate this task once we have that.

**Comment by zufanka** [comment]
I personally don't like giving random application access to any of my social media accounts :)

But yes, that would make the hassle way less. And if we have public networks that the user can play with without having to go through building a query, it also might be ok to just leave creation of networks behind a login.

@chutko what do you think?

**Comment by chutko** [comment]
yes, third party login would make it easier, but its still a registration - an obstacle we can let the user avoid. 

having public networks functionality will definitely come useful in the future in many ways - sharing the tool, spreading possibly influencial information more easily, making selections of interesting networks for the public to explore. so its good not to have the way blocked once this potential is fully needed. 

i like to imagine a future political cause which could elvis be part of by helping spreading information and winning a small battle against corruption. when internet gets flooded with some topic, its crucial that information (=arguments) are easily spread and accessible so more people can get involved and the topic is discussed more intensively therefore possibly bringing some solution to the problem. 

so, not asking for registration to access information from elvis seems like a small thing - but as you can see, lives of our children depend on it

**Comment by chutko** [comment]
what does this feature involve? what subissues shall we create to make it fully running?
feel free to check what is already solved and add what is missing

- [ ] backend stuff i have no idea about
- [ ] checkbox in the network settings (and/or in the sharing dialog) to set the network public
- [ ] create 2-3 example public networks
- [ ] create landing page with displayed public networks - route, design and frontend
- [ ] remove system restriction in the query builder (cannot click Visualise button if not logged in)
- [ ] add homepage call to action button linking to example networks
- [ ] fix #67 (this topic is a bit login related, so why not?)
- [ ] fix #195  (same)

**Comment by zufanka** [comment]
@nightsh do we already have any (stable) public network?

**Comment by zufanka** [comment]
Could we tackle this issue in the near future? @nightsh

**Comment by georgiana-b** [comment]
@chutko 

> i like to imagine a future political cause which could elvis be part of by helping spreading the information resulting into winning a small battle against corruption. when internet gets flooded with some topic, its crucial that information (=arguments) are easily spread and accessible so more people can get involved and the topic is discussed more intensively therefore possibly bringing some solution to the problem.

That is touching 😄 We actually make that possible already by allowing free access *to view* the networks. So, if somebody shares the link to a network, anybody can see it.

This issue is about people being allowed *to create* networks without logging in.
There are some reasons against this:
- *creating* a network is expensive for the server
    - we store the networks in the db and we should encourage people to create accounts and *retrieve* the network they already created (much cheaper) instead of *creating* the network every time
    - anybody evil or eager can kill our servers by creating many consuming networks over few seconds
- if we allow networks that are not attached to any account we have to manage them ourselves:
    - how long do nobody's networks live?
    - who can change them?
    - who can delete them?
    - on what criteria do we delete them to avoid running out of server space?

In [this comment](https://github.com/tenders-exposed/elvis-ember/issues/167#issuecomment-304064730) I tried to address some of these issues. 

However, your comments gave me a better idea: we can add a public networks collection. So, a user can choose to include one of his/her networks in the public collection which means anybody who comes on Elvis can see it. This public networks collection will probably look similar with My projects page.
Benefits of the public networks collection:
- give newcomers ability to see networks without making an account
- the networks on that page will be somehow curated since people found them worthy of sharing, not just random trials
- those networks will still be managed by their owner
- we have to add very little functionality to make it happen since these will be just normal networks that can already be shared and seen by anyone with the link; we just gather some links on the collection page

What do you all think?

**Comment by chutko** [comment]
@georgiana-b 

> how long do nobody's networks live?
for now lets set it to 2 weeks for instance. if there is no activity for this period (=0 views), network is deleted. in the future, this could be the network owner's option 

> who can change them? who can delete them?
this also can be an option with checkbox in the public network settings - "anybody" or "just me". in the future, we can have option to share network with specific users and (dis)allow them to manipulate with network

in the very first version of this whole feature, we can restrict everything to minimum and add other options later:
1. if the network is private - only the owner can read, update and delete (RUD) this network
2. if the network is public - owner can RUD and others can view (OR anybody can RUD this network?)

**Comment by georgiana-b** [comment]
@chutko 

> this also can be an option with checkbox in the public network settings - "anybody" or "just me"

I think there's some confusion so let's clarify. This issue is about the following scenario:
A person lands on our website and makes a network without logging in. This means we now have a network in the db that is not attached to any user account. The person wanders at the network but then leaves our website without ever making an account.

So, the options you mention: "anybody" or "just me" are not applicable because there is no "me" if the person doesn't have an account.

**Comment by georgiana-b** [comment]
My proposal is: we shouldn't even allow this scenario, a user should not be able to create a network without an account. 

However, to give a person that is not yet registered an idea of how networks looks like, we crate a page that looks like "My projects" page, where we list networks that other users chose to share.
This means my 2 options are:

1. if the network is private  - only the owner can *update* and *delete* this network (UD), anybody can *see*/*read* it if they have the link to the network; The link to the network is pretty hard to guess though so if somebody has it it's probably because the owner gave it to them
2. if the network is public - only the owner can *update* and *delete* this network (UD), anybody can *see*/*read* if they have the link; So, same as above, but we list these on a public page where people can see them without having to guess the link. Practically, the owner decides to share the link to the network with anybody who comes on Elvis

From your 2 options I don't understand how a stranger with no account or link to a network can get to  see one. Pls explain.

**Comment by zufanka** [comment]
How about we just provide a dummy user that people without an account can use @chutko ?

It has multiple advantages:
- no need to think up new rules for anonymous users
- there will be already random networks created in "My projects" (can be pro and a con at the same time tho). W can set a limit to the saved networks of the dummy user then (max 50 or so).

I would not complicate it now with private / public networks. All networks should be publicly accessible if provided the URL. Otherwise they are stored in "My projects". If there is a need for the private/public in the future we can add it.

Updating a network is still an issue on it's own. Afaik it's not possible now, so we should also see whether we want / need this to happen before the end of the year (and the dni grant) or not.

@georgiana-b I can't load [this network](http://tenders.exposed/network/59ac47382dcb560cd8000001). It's stuck on Stabilizing 0% :/

**Comment by zufanka** [comment]
@chutko [template networks @ discourse](https://talk.tenders.exposed/t/saving-template-networks-which-can-be-just-duplicated/99)

**Comment by georgiana-b** [comment]
Ok, I see you both really want to let anonymous users create networks 😺 
A mock account would be troublesome technically & it doesn't  help us avoid the questions so please object or contribute to any of the answers:

1. Who will see the userless networks? 
    Anyone.
2. Who will be able to edit them?
    Nobody.
3. Who will be able to delete them?
    Nobody but us in the background.
4. When will we delete them?
    If they haven't been used for 2 weeks, always ensuring we have at least 10.
5. Where will these networks be found?
    We will make a public page that looks like "My projects" that will list them.

**Comment by zufanka** [comment]
> Ok, I see you both really want to let anonymous users create networks 😺
I thought this is something we have agreed on some time ago ... ?

> A mock account would be troublesome technically 
we have one right now, the data harvest account :)

1. anyone who is logged in with the user
2. idem
3. idem, just anarchy \m/
4. we can set s max saved for that account or wiped every x days
5. the same My projects as any other account

Just to be clear: I am trying to find the easiest solution here, not necessarily the best one. We can still think of another one at a later stage.

The best solution is in my opinion to let users draw networks without logging in that won't be saved anywhere and will disappear once you close the browser window. That way there would no need for 1 2 3 4 5 to be considered at all.

**Comment by georgiana-b** [comment]
@zufanka 

> The best solution is in my opinion to let users draw networks without logging in that won't be saved anywhere and will disappear once you close the browser window.

This sounds pretty good but I think it was dogged because it sounds fringe to detect when the browser window is closed.

@nightsh  Is ☝️   true? Could you trigger a DELETE network event when the  user moves away from the route `network/:id` or when the browser window is closed?

**Comment by zufanka** [comment]
Can't / doesn't the network exist in the browser after it is drawn? (just poking in the dark here)

**Comment by georgiana-b** [comment]
I'm laughing at the fact you're poking in the dark not at the proposal which could be a good one.

**Comment by chutko** [comment]
i actually like georgiana's idea that we will not have any anonymous accounts and just let anybody to browse the public networks list. its currently the simplest solution 

the only thing which goes against it in my head is - if people want to see some network, they want to see something they know and can refer to - so i am from Slovakia and the only kind of network which i can relate to and am able to properly assess is procurement in Slovakia in fields of health and road construction (because those are the ones known to be the most corrupted). and if i will not find these networks in the list of public networks, the system will probably be not interesting for me (or i will have to sign up which we would like to avoid). so this problem is a big one for me. otherwise i dont really care. 

so again, im gonna try to solve the problems georgiana pointed to at the beginning:

> how long do nobody's networks live?

_deleted_ right after the session ends or _saved_ if user creates an account

> to detect when the browser window is closed

google "_session variables_" (PHP term) - there must be some equivalent for it 

> who can change them? who can delete them?

if the network belongs to signed user, user can change and delete. if it is an _anonymous network_, no manipulation (clustering, naming, embedding etc) is available (or requires registration). 

> on what criteria do we delete them to avoid running out of server space?

_anonymous networks_ are deleted after browser session ends and other networks (registered users' networks) are stored.

**Comment by georgiana-b** [comment]
So the preferred option is to let users make networks that are deleted after they move away from the network's route or close the browser window. 

This needs some research from @nightsh  or @ca1yps0  to find out if it's possible to determine when the user closed the browser window that contained the network.

If it's possible we go with it, if not we'll choose one of the other options we mentioned here.

**Comment by nightsh** [comment]
So the current way to do it is:

* user makes a network request
* network gets created
  * then saved
  * then returned as a payload to the browser
* network gets displayed with all the bells and whistles we have

---

Why can't we do exactly the above for unauthenticated requests, minus the saving part? We can take care of notifications telling people we won't save anything unless they log in.

This seems the least intrusive way to me, also the easiest to do with what we have now. Am I missing anything? Does it make sense? @georgiana-b

**Comment by georgiana-b** [comment]
HTTP doesn't agree with this since there is no network unless you create it and there is no creating without saving it.
To return the created network on POST without saving it to the database is a evil thing to do as an API.

 @nightsh, can you describe the request you would like to make in order to achieve without breaking HTTP rules?

**Comment by nightsh** [comment]
Same request. If there are no valid auth headers present, then don't save, or rather don't keep the newly created network.
As for rules and good practice... that pretty much flies out the window if we do this anyway.

If we want to be make a good case of development practices, I'd say we should save them and then run a garbage collector-like process over them every now and then to remove what's older than the maximum anonymous network lifetime. This wouldn't violate anything AFAICT, but I wouldn't cluster / rename / any r/w ops on them unless logged in and fully authorized to do any request.

**Comment by nightsh** [comment]
Ping @georgiana-b, WDYT?

**Comment by nightsh** [comment]
We could also detect when a use closes the window / navigates away from the network. That's possibly faulty though, but worth experimenting with anyway.  It's not about the technology itself, but more about modern browsers attempting to fight some JS callbacks in an effort to reduce spammy behaviour of some sites (imagine websites that throw alerts and dialog windows when you close them 😱 )
I would like to have a backup plan, such as time-based expiration, in case this fails.

**Comment by georgiana-b** [comment]
In case this fails we will create them normally and delete them with a background task after 2 weeks.

**Comment by zufanka** [comment]
@nightsh can we delete networks after the browser is closed??

**Comment by victor-nitu** [comment]
It's a bit unreliable to do so. I'd go with time based expiration of anon. networks, if that complies with the requirements. Agreeing with @georgiana-b on this :)

**Comment by zufanka** [comment]
@nightsh I think 2 weeks is long, we can also do that once a couple of days.
Even more incentive to create an account :)

**Comment by zufanka** [comment]
@nightsh this is anyway for now blocked by the new backend, or not?

**Comment by georgiana-b** [comment]
> this is anyway for now blocked by the new backend, or not?

Yes.
I synthesized this discussion into specs in tenders-exposed/elvis-backend-py#10. You are all welcomed to go take a look at them and make any objections.

**Comment by zufanka** [comment]
ping @georgiana-b is there an issue for this in the new backend?

**Comment by georgiana-b** [comment]
Yes, it's included in tenders-exposed/elvis-backend-node#7

**Comment by georgiana-b** [comment]
@nightsh Currently the frontend welcomes users with a `Create an account` button. That should be turned into `Create a visualization` like it was before since people will be able to create networks before having an account. 

Do you think we should open a new issue for that or this one will do?
